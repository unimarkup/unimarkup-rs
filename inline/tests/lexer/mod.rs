use std::panic;

use libtest_mimic::Trial;
use unimarkup_commons::test_runner::{self, snap_test_runner::SnapTestRunner};
use unimarkup_inline::Tokenize;

use crate::snapshot::Snapshot;

mod snapshot;

pub fn test_lexer_snapshots() -> Vec<Trial> {
    let tests_path = unimarkup_commons::crate_tests_path!();
    let test_cases = test_runner::collect_tests(
        tests_path.join("spec/markup"),
        tests_path.join("spec/snapshots/lexer"),
        "markup",
    );

    let mut test_runs = Vec::with_capacity(test_cases.len());
    for case in test_cases {
        let test_name = format!("{}::{}", module_path!(), case.test.name.as_str());

        let test_run = move || {
            panic::catch_unwind(|| run_test_case(case)).map_err(|err| {
                let panic_msg = err
                    .downcast_ref::<&str>()
                    .unwrap_or(&"Panic message not available");

                format!("Test case panicked: {}", panic_msg).into()
            })
        };

        test_runs.push(Trial::test(test_name, test_run));
    }

    test_runs
}

fn run_test_case(case: test_runner::test_file::TestCase) {
    let mut symbols = unimarkup_commons::scanner::scan_str(&case.test.input);
    symbols.pop(); // Remove EOI symbol for test cases

    let runner = SnapTestRunner::with_fn(&case.test.name, &symbols, |symbols| {
        Snapshot::snap((case.test.input.as_ref(), symbols.tokens()))
    })
    .with_info(format!(
        "Test '{}' from '{}'",
        case.test.name, case.file_path
    ));

    unimarkup_commons::run_snap_test!(runner, &case.out_path);
}
