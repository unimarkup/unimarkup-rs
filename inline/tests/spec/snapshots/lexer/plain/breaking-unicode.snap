---
source: inline/tests/lexer/mod.rs
info: "Test 'breaking-unicode' from 'markup/plain.yml'"
---
The character 𝛂 broke some stuff.
The
^^^ -> Plain @ (1:1)->(1:4)
    
   ^ -> Whitespace @ (1:4)->(1:5)
    character
    ^^^^^^^^^ -> Plain @ (1:5)->(1:14)
              
             ^ -> Whitespace @ (1:14)->(1:15)
              𝛂
              ^ -> Plain @ (1:15)->(1:16)
                
               ^ -> Whitespace @ (1:16)->(1:17)
                broke
                ^^^^^ -> Plain @ (1:17)->(1:22)
                      
                     ^ -> Whitespace @ (1:22)->(1:23)
                      some
                      ^^^^ -> Plain @ (1:23)->(1:27)
                           
                          ^ -> Whitespace @ (1:27)->(1:28)
                           stuff.
                           ^^^^^^ -> Plain @ (1:28)->(1:34)
                                 ⏎
                                 ^ -> Newline @ (1:34)->(2:1)

---
With input:

The character 𝛂 broke some stuff.


